File: Python/318__Maximum_Product_of_Word_Lengths.py
class Solution(object):
    def maxProduct(self, words):
        """
        :type words: List[str]
        :rtype: int
        """
        result = 0
        for i in range(len(words)):
            for j in range(i + 1, len(words)):
                if not (set(words[i]) & set(words[j])):
                    result = max(result, len(words[i]) * len(words[j]))
        return result

# Or Bit Manipulation, much faster!

class Solution(object):
    def maxProduct(self, words):
        """
        :type words: List[str]
        :rtype: int
        """
        # 26-bits: store distinct letters
        values = [0] * len(words)
        result = 0
        for i in range(len(words)):
            for char in words[i]:
                values[i] |= 1 << (ord(char) - ord('a'))
            for j in range(i):
                if values[i] & values[j] == 0:
                    result = max(result, len(words[i]) * len(words[j]))
        return result

File: Java/318__Maximum_Product_of_Word_Lengths.java
class Solution {
    public int maxProduct(String[] words) {
        int len = words.length;
		int [] words_bit = new int[len];
		int [] lenList = new int[len];
		int i = 0;
		for (String word : words) {
			for (char c : word.toCharArray())
				words_bit[i] |= (1 << (int) (c - 'a'));
			lenList[i] = word.length();
			i++;
		}
		int result = 0;
		for (i = 0; i < len; i++)
			for (int j = i + 1; j < len; j++)
				if ((words_bit[i] & words_bit[j]) == 0)
					result = Math.max(result, lenList[i] * lenList[j]);
		return result;
    }
}
